docs/ParameterValidation.md

```diff
@@ -0,0 +1,174 @@
+- Valid input with no results returns 200 with empty array
+- Invalid input returns 400 text/plain with Invalid error response
+
+### Single Reads
+
+- Valid single reads returns 200 with JSON document
+- Valid single reads with no results returns 404 with No {entity} found error response
+- Invalid single reads like /api/movies/tt1234 , /api/actorid/nm1234 returns 400 with Invalid error response
+
+### Multiple Error Scenarios
+
+- Fail on first error
+
+## Movies
+
+### API /api/movies
+
+### Movies Query Parameters
+
+- Name: query (search)
+- Description: This a search query string on the movie title or actor name field
+- Type: string
+- Parameter Validation:
+  - Valid input: query length >= 3 non-whitespace characters
+    - Status: 200
+    - Content-Type: application/json
+    - Response Body: JSON Array of movies filtered by search query on title or actor name field as [{}, {},..], empty array when no results
+  - Invalid input: query length < 3 characters
+    - Status: 400
+    - Content-Type: text/plain
+    - Response Body: Invalid query parameter
+
+- Name: year
+- Type: integer
+- Parameter Validation:
+  - Valid input range: [1874 , currentYear+5]
+    - Status: 200
+    - Content-Type: application/json
+    - Response Body: JSON Array of movies filtered by year as [{}, {},..], empty array when no results
+  - Invalid input - input that does not parse or is out of range
+    - Status: 400
+    - Content-Type: text/plain
+    - Response Body: Invalid year parameter
+
+- Name: rating
+- Type: double
+- Parameter Validation:
+  - Valid input range: [0.0, 10.0]
+  - Round off more than single decimal to a single decimal
+    - Example: round off 7.253896 to 7.3
+    - Status: 200
+    - Content-Type: application/json
+    - Response Body: JSON Array of movies filtered by rating as [{}, {},..], empty array when no results, where movie.rating >= rating
+  - Invalid input: input that does not parse or is out of range
+    - Status: 400
+    - Content-Type: text/plain
+    - Response Body: Invalid rating parameter
+
+- Name: actorId
+- Type: string
+- Parameter Validation
+  - Valid input : well formed actorId rule - starts with 'nm' followed by 5-9 digits, parsed as an int
+    - Example: 'nm1265067'
+    - Status : 200
+    - Content-Type: application/json
+    - Response Body: JSON Array of movies filtered by actorid as [{}, {},..], empty array, when no results, where actorId was in movie
+  - Invalid input: values not conforming to well formed actorid rule.  
+    - Example: 'nm1234', 'ab1234'
+    - Status : 400
+    - Content-Type: text/plain
+    - Response Body: Invalid actorid parameter
+
+- Name: genre
+- Type: string
+- Parameter Validation:
+  - Valid input: genre length [3, 20]
+    - Example: 'War' , 'Documentary'
+    - Status: 200
+    - Content-Type: application/json
+    - Response Body: JSON Array of movies filtered by genre as [{}, {},..], empty array when no results,  where movie.genres contains genre
```

Comment:
This is currently case insensitive - should we call that out?

Reply:
this is addressed